node/src/components/sync_leaper/leap_state.rs

```diff
@@ -0,0 +1,156 @@
+use std::fmt::{Display, Formatter};
+
+use datasize::DataSize;
+
+use crate::types::{NodeId, SyncLeap, SyncLeapIdentifier};
+
+use super::LeapActivityError;
+
+#[derive(Debug, DataSize)]
+pub(crate) enum LeapState {
+    Idle,
+    Awaiting {
+        sync_leap_identifier: SyncLeapIdentifier,
+        in_flight: usize,
+    },
+    Received {
+        best_available: Box<SyncLeap>,
+        from_peers: Vec<NodeId>,
+        in_flight: usize,
+    },
+    Failed {
+        sync_leap_identifier: SyncLeapIdentifier,
+        error: LeapActivityError,
+        from_peers: Vec<NodeId>,
+        in_flight: usize,
+    },
+}
+
+impl Display for LeapState {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            LeapState::Idle => {
+                write!(f, "Idle")
+            }
+            LeapState::Awaiting {
+                sync_leap_identifier,
+                in_flight,
+            } => {
+                write!(
+                    f,
+                    "Awaiting {} responses for {}",
+                    in_flight,
+                    sync_leap_identifier.block_hash(),
+                )
+            }
+            LeapState::Received {
+                best_available,
+                from_peers,
+                in_flight,
+            } => {
+                write!(
+                    f,
+                    "Received {} from {} peers, awaiting {} responses",
+                    best_available.highest_block_hash(),
+                    from_peers.len(),
+                    in_flight
+                )
+            }
+            LeapState::Failed {
+                sync_leap_identifier,
+                error,
+                ..
+            } => {
+                write!(
+                    f,
+                    "Failed leap for {} {}",
+                    sync_leap_identifier.block_hash(),
+                    error
+                )
+            }
+        }
+    }
+}
+
+impl LeapState {
+    pub(super) fn in_flight(&self) -> usize {
+        match self {
+            LeapState::Idle => 0,
+            LeapState::Awaiting { in_flight, .. }
+            | LeapState::Received { in_flight, .. }
+            | LeapState::Failed { in_flight, .. } => *in_flight,
+        }
+    }
+
+    pub(super) fn active(&self) -> bool {
+        self.in_flight() > 0
```

Comment:
nit: it would be better to do this with an explicit exhaustive match on `LeapState` so that any further changes to the enum will make us revisit this function.

Reply:
Agreed, but I'm not touching this right now to avoid accidental regressions.